src/js/node/http2.ts

```diff
@@ -350,348 +1297,1498 @@ const constants = {
+          this.emit("aborted");
+        }
 
-  get sentTrailers() {
-    return this.#sentTrailers;
+        // at this state destroyed will be true but we need to close the writable side
+        this._writableState.destroyed = false;
+        this.end();
+        // we now restore the destroyed flag
+        this._writableState.destroyed = true;
+      }
+
+      const session = this[bunHTTP2Session];
+      assertSession(session);
+
+      let rstCode = this.rstCode;
+      if (!rstCode) {
+        if (err != null) {
+          if (err.code === "ABORT_ERR") {
+            // Enables using AbortController to cancel requests with RST code 8.
+            rstCode = NGHTTP2_CANCEL;
+          } else {
+            rstCode = NGHTTP2_INTERNAL_ERROR;
+          }
+        } else {
+          rstCode = this.rstCode = 0;
+        }
+      }
+
+      if (this.writableFinished) {
+        markStreamClosed(this);
+
+        session[bunHTTP2Native]?.rstStream(this.#id, rstCode);
+        this[bunHTTP2Session] = null;
+      } else {
+        this.once("finish", Http2Stream.#rstStream);
+      }
+    } else {
+      this[bunHTTP2Session] = null;
+    }
+
+    callback(err);
   }
 
-  sendTrailers(headers) {
+  _final(callback) {
+    const status = this[bunHTTP2StreamStatus];
+
+    if ((status & StreamState.WritableClosed) !== 0 || (status & StreamState.Closed) !== 0) {
+      callback();
+      this[bunHTTP2StreamStatus] |= StreamState.FinalCalled;
+    } else {
+      this[bunHTTP2StreamFinal] = callback;
+    }
+  }
+
+  _read(size) {
+    const queue = this[bunHTTP2StreamReadQueue];
+    let chunk;
+    while ((chunk = queue.peek())) {
+      if (!this.push(chunk)) {
+        queue.shift();
+        return;
+      }
+      queue.shift();
+    }
+  }
+
+  end(chunk, encoding, callback) {
+    const status = this[bunHTTP2StreamStatus];
+
+    if ((status & StreamState.EndedCalled) !== 0) {
+      typeof callback == "function" && callback();
+      return;
+    }
+    if (!chunk) {
+      chunk = Buffer.alloc(0);
+    }
+    this[bunHTTP2StreamStatus] = status | StreamState.EndedCalled;
+    return super.end(chunk, encoding, callback);
+  }
+
+  _writev(data, callback) {
     const session = this[bunHTTP2Session];
-    assertSession(session);
+    if (session) {
+      const native = session[bunHTTP2Native];
+      if (native) {
+        const allBuffers = data.allBuffers;
+        let chunks;
+        chunks = data;
+        if (allBuffers) {
+          for (let i = 0; i < data.length; i++) {
+            data[i] = data[i].chunk;
+          }
+        } else {
+          for (let i = 0; i < data.length; i++) {
+            const { chunk, encoding } = data[i];
+            if (typeof chunk === "string") {
+              data[i] = Buffer.from(chunk, encoding);
+            } else {
+              data[i] = chunk;
+            }
+          }
+        }
+        const chunk = Buffer.concat(chunks || []);
+        native.writeStream(
```

Comment:
What happened to adding support for encoding?

Reply:
We are using Buffer.concat so we just use `Buffer.from(chunk, encoding);` the optimized version would be in a followup (all in Zig)